middleware/observation/utils.py

```diff
@@ -0,0 +1,434 @@
+            return None
+        return {
+            "systolic": observation.systolic.value,
+            "diastolic": observation.diastolic.value,
+        }
+    else:
+        return observation.value
+
+
+def get_stored_observations():
+    observations = cache.get(settings.REDIS_OBSERVATIONS_KEY)
+    if observations is None:
+        observations = []
+        cache.set(settings.REDIS_OBSERVATIONS_KEY, observations)
+
+    return ObservationList.model_validate(observations).root
+
+
+def update_stored_observations(observation_list: List[Observation]):
+    # observations = cache.get(settings.REDIS_OBSERVATIONS_KEY)
+    # if observations is None:
+    # observations = []
+    # observations.extend(observation_list)
+    cache.set(settings.REDIS_OBSERVATIONS_KEY, observation_list)
+
+
+def extract_datetime(key):
+    # Split the string to get the timestamp part
+    timestamp_str = key.split("_", 1)[1]
+    # Convert the string to a datetime object
+    return datetime.fromisoformat(timestamp_str)
+
+
+def get_static_observations(device_id: DeviceID):
+    observation_keys = cache.keys(f"{settings.REDIS_OBSERVATIONS_KEY}*")
+
+    sorted_keys = sorted(observation_keys, key=extract_datetime)
+    logger.info("first key is %s", sorted_keys[0])
+    logger.info("last key is %s", sorted_keys[-1])
+    observations = []
+    for key in sorted_keys:
+        observations.extend(cache.get(key))
+    stale_time = now() - timedelta(minutes=settings.UPDATE_INTERVAL)
+
+    # last one hour data matching the device id
+    valid_observations: List[Observation] = []
+    if not observations:
+        logger.info(" No observations for device id : %s stored in redis ", device_id)
+        return None
+    for observation in observations:
+        parsed_observation = Observation.model_validate(observation)
+
+        if (
+            parsed_observation.taken_at > stale_time
+        ) and parsed_observation.device_id == device_id:
+            valid_observations.append(parsed_observation)
+
+    if not valid_observations:
+        logger.info(
+            " No observations Valid observations for device id : %s stored in redis ",
+            device_id,
+        )
+        return None
+    return generate_static_observations(observation_list=valid_observations)
+
+
+def generate_static_observations(observation_list: List[Observation]):
+    observations_dict = {}
+    for observation in observation_list:
+        observation_type = observation.observation_id
+        if observation_type in observations_dict:
+            observations_dict[observation_type].append(observation)
+        else:
+            observations_dict[observation_type] = [observation]
+
+    return StaticObservation(
+        observations=observations_dict, last_updated=observation_list[-1].taken_at
+    )
+
+
+def get_data_for_s3_dump():
+    observations = cache.get(settings.REDIS_OBSERVATIONS_KEY)
+    current_time = datetime.now()
```

Comment:
we switched to storing each observation request separately, have you not updated this?